charms/metallb-controller/metadata.yaml

```diff
@@ -21,3 +21,4 @@ resources:
   metallb-controller-image:
     type: oci-image
     description: upstream docker image for metallb-controller
+    upstream-source: 'metallb/controller:v0.9.3'
```

Comment:
Inspired by your conversation with Ken, in the CoreDNS charm I'm working on I used the following in the test to use this field to drive the test as well, so that it only needs to be updated in one spot:

```yaml
      run: |
        upstream_image=$(yq r metadata.yaml 'resources.coredns-image.upstream-source')
        sudo juju deploy ./coredns.charm --resource coredns-image=$upstream_image --config forward=8.8.8.8
```

Reply:
ah not a bad idea!